src/SAML2/Assertion.php

```diff
@@ -1243,24 +1143,84 @@ public function setAuthenticatingAuthority(array $authenticatingAuthority): void
 
     /**
      * Retrieve all attributes.
-     *
-     * @return array All attributes, as an associative array.
+     * 
+     * @return \SAML2\XML\saml\Attribute[] All attributes, as an associative array.
      */
     public function getAttributes(): array
     {
         return $this->attributes;
     }
 
-
+    
+    /**
+     * Retrieve all attribute values as an associative array
+     *
+     * @return array of attribute values
+     */
+    public function getAttributeValues(): array
+    {
+        $attributes = [];
+        
+        foreach ($this->attributes as $attributeObj) {
+            $attributeName = $attributeObj->getName();
+            $attributes[$attributeName] = [];
+            
+            foreach ($attributeObj->getAttributeValue() as $attributeValue) {
+                $value = $attributeValue->getValue();
+                
+                // need to determine if we should return a NameID
+                if ($attributeName === Constants::EPTI_URN_MACE || $attributeName === Constants::EPTI_URN_OID) {
+                    $nameId = null;
+                    if ($value instanceof DOMNodeList) {
+                        foreach ($value as $node) {
+                            /** @var \DOMNode $node */
+                            if ($node->nodeType !== XML_TEXT_NODE && $node->localName === 'NameID' && $node->namespaceURI === Constants::NS_SAML) {
+                                $nameId = new NameID($node);
+                            }
+                        }
+                    } else {
+                        /* Fall back for legacy IdPs sending string value (e.g. SSP < 1.15) */
+                        Utils::getContainer()->getLogger()->warning(
+                            sprintf("Attribute %s (EPTI) value %d is not an XML NameId", $attributeName, $value)
+                        );
+                        $nameId = new NameID();
+                        $nameId->setValue($value);                    
+                    }  
+                    if ($nameId !== null) {
+                        $value = $nameId;
+                    }
+                }            
+                $attributes[$attributeName][] = $value;
+            }
+        }
+        return $attributes;
+    }
+    
+    
     /**
      * Replace all attributes.
      *
-     * @param array $attributes All new attributes, as an associative array.
+     * @param \SAML2\XML\saml\Attribute[] $attributes All new attributes, as an associative array.
      * @return void
      */
     public function setAttributes(array $attributes): void
     {
-        $this->attributes = $attributes;
+        foreach ($attributes as $name => $value) {
+            if ($value instanceof Attribute) {
+                $this->attributes[$name] = $value;
+            } else {
```

Comment:
Why do you want to keep BC-supprt for non-Attribute objects?

Reply:
Again, I’m a fan of BC. At least for period of time while it’s made clear to folks who  are using the library that things are going to change and they should be doing things the new way. In that respect, I should have added some logging to warn about passing non-attribute objects. Again, you own this code so if you want to remove it, feel free. However I think that it would be nice to add an array typehint to ensure that the array only contains attribute objects like:

public function setAttributes(Attribute ...$attributes)

However this won’t work on PHP versions < 5.6 and I’m pretty sure we need to support 5.4

Reply:
That would allow a variable amount of Attribute-objects passed as paramaters, not an array of Attribute-objects, or am I missing something? We don't have to support PHP5.x anymore.. v4.x of this library actually requires >=7.2

I'll discuss the BC with Jaime.. I think he's away for the holidays, so it may take a bit..

Reply:
Oh oops. Well I guess there is no way I’m PHP to force an array to only contain a certain type of element unless I’m missing something. So having that code was merely a way to ensure that only Attribute objects were kept internally